### ISSUE TITLE
hab abandons services after SIGHUP

### ISSUE THREAD

**Issue by jaym** [habitat-sh/habitat#5687]
Trying out the new feature where SIGHUPing hab-sup causes it to restart, I'm finding that it somewhat abandons services after the SIGHUP. Below is a studio session where I have a spec file for redis. I start by creating a symlink called supy, which points to the hab-sup I want to run. I start the supervisor using the `HAB_SUP_BINARY` environment variable pointing at that. After everything comes up, I overwrite the supy symlink with a newer version of habitat and do a SIGHUP against supy. Things reload correctly, however doing a `sup-term` does not kill redis

```
[118][default:/src:0]# ps aux | grep redis
root     25622  0.0  0.0   9120   476 ?        S+   20:47   0:00 grep --color=auto redis
[119][default:/src:0]# ps aux | grep launcher
1000     10357  0.0  0.0 324384  5652 ?        Ssl  Sep24   0:00 /usr/lib/at-spi-bus-launcher
root     25749  0.0  0.0   9124   472 ?        S+   20:47   0:00 grep --color=auto launcher
[120][default:/src:0]# hab svc status
✗✗✗
✗✗✗ Connection refused (os error 111)
✗✗✗
[121][default:/src:1]# ln -sf /hab/pkgs/core/hab-sup/0.62.0/20180831211729/bin/hab-sup supy
[122][default:/src:0]# readlink supy
/hab/pkgs/core/hab-sup/0.62.0/20180831211729/bin/hab-sup
[123][default:/src:0]# HAB_SUP_BINARY=/src/supy sup-run
--> Launching the Habitat Supervisor in the background...
    Running: hab sup run 
[1] 26198
    * Use 'hab svc start' & 'hab svc stop' to start and stop services
    * Use 'sup-log' to tail the Supervisor's output (Ctrl+c to stop)
    * Use 'sup-term' to terminate the Supervisor
    * To pass custom arguments to run the Supervisor, export
      'HAB_STUDIO_SUP' with the arguments before running
      'hab studio enter'.

[124][default:/src:0]# hab svc status
package                           type        state  elapsed (s)  pid    group
core/redis/4.0.10/20180801003001  standalone  up     5            26319  redis.default
[125][default:/src:0]# ln -sf /hab/pkgs/core/hab-sup/0.63.0/20180914030447/bin/hab-sup supy
[126][default:/src:0]# ps aux | grep hab-sup
root     26677  0.0  0.0   9120   472 ?        S+   20:49   0:00 grep --color=auto hab-sup
[127][default:/src:0]# ps aux | grep supy   
root     26202  0.1  0.0 415148 20952 ?        Sl   20:48   0:00 /src/supy run
root     26729  0.0  0.0   9120   472 ?        S+   20:49   0:00 grep --color=auto supy
[128][default:/src:0]# kill -s SIGHUP 26202
[129][default:/src:0]# sl
--> Tailing the Habitat Supervisor's output (use 'Ctrl+c' to stop)
redis.default(O): 26319:M 28 Sep 20:48:21.188 * Running mode=standalone, port=6379.
redis.default(O): 26319:M 28 Sep 20:48:21.188 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
redis.default(O): 26319:M 28 Sep 20:48:21.188 # Server initialized
redis.default(O): 26319:M 28 Sep 20:48:21.188 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
redis.default(O): 26319:M 28 Sep 20:48:21.188 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
redis.default(O): 26319:M 28 Sep 20:48:21.188 * DB loaded from disk: 0.000 seconds
redis.default(O): 26319:M 28 Sep 20:48:21.188 * Ready to accept connections
supy(MR): Supervisor shutting down for signal
supy(MR): Gracefully departing from butterfly network.
ERROR 2018-09-28T20:49:36Z: habitat_launcher::server: Supervisor exited; will restart in 5 seconds...
supy(MR): Supervisor Member-ID aba57b2a8dbc471c973c1b98e3050769
supy(MR): Starting core/redis
redis.default(UCW): Watching user.toml
supy(MR): Starting gossip-listener on 0.0.0.0:9638
supy(MR): Starting ctl-gateway on 127.0.0.1:9632
supy(MR): Starting http-gateway on 0.0.0.0:9631
redis.default(HK): Hooks compiled
supy(SR): Reattached to redis.default
^C
[130][default:/src:130]# sup-term
--> Killing Habitat Supervisor running in the background...
    Supervisor killed.
[131][default:/src:0]# sl
--> Tailing the Habitat Supervisor's output (use 'Ctrl+c' to stop)
supy(MR): Gracefully departing from butterfly network.
ERROR 2018-09-28T20:49:36Z: habitat_launcher::server: Supervisor exited; will restart in 5 seconds...
supy(MR): Supervisor Member-ID aba57b2a8dbc471c973c1b98e3050769
supy(MR): Starting core/redis
redis.default(UCW): Watching user.toml
supy(MR): Starting gossip-listener on 0.0.0.0:9638
supy(MR): Starting ctl-gateway on 127.0.0.1:9632
supy(MR): Starting http-gateway on 0.0.0.0:9631
redis.default(HK): Hooks compiled
supy(SR): Reattached to redis.default
supy(MR): Gracefully departing from butterfly network.
hab-launch(SV): Hasta la vista, services.
ERROR 2018-09-28T20:49:59Z: hab_launch: Launcher exiting with code 0
^C[1]+  Done                    hab sup run "$@" > /hab/sup/default/sup.log 2>&1

[132][default:/src:130]# ps aux | grep redis
hab      26319  0.0  0.0  42288  3852 ?        Sl   20:48   0:00 redis-server *:6379
```

The same thing happens when I try going from 0.64.0 to 0.64.1

**Comment by stevendanna** [comment]
Note, I can reproduce this behavior without any of the upgrading or symlinking involved:

```
[62][default:/src:0]# ps fauxw | grep [h]ab-
root       512  0.0  0.0  48056  6040 pts/0    Sl   23:26   0:00  \_ /hab/pkgs/core/hab-launcher/8479/20180917170345/bin/hab-launch run
root       516  0.2  0.1 209600 15276 pts/0    Sl   23:26   0:00  |   \_ /hab/pkgs/core/hab-sup/0.64.1/20180928012737/bin/hab-sup run
[63][default:/src:0]# ps fauxw | grep [r]edis
hab        569  0.2  0.0  42276  3764 pts/0    Sl   23:26   0:00  |   \_ redis-server *:6379
[64][default:/src:0]# hab svc status
package                           type        desired  state  elapsed (s)  pid  group
core/redis/4.0.10/20180801003001  standalone  up       up     62           569  redis.default
[65][default:/src:0]# kill -HUP 516
[66][default:/src:0]# hab-sup(MR): Supervisor shutting down for signal
ERROR 2018-09-28T23:27:16Z: habitat_launcher::server: Supervisor exited; will restart in 5 seconds...
hab-sup(MR): core/hab-sup (core/hab-sup/0.64.1/20180928012737)
hab-sup(MR): Supervisor Member-ID 76bdcd19e5a0421fac26135e3cd893fd
hab-sup(MR): Starting core/redis (core/redis/4.0.10/20180801003001)
redis.default(UCW): Watching user.toml
hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
redis.default(HK): Hooks compiled
hab-sup(SR): Reattached to redis.default

[66][default:/src:0]# kill 512
[67][default:/src:0]# hab-sup(MR): Gracefully departing from butterfly network.
hab-launch(SV): Hasta la vista, services.
ERROR 2018-09-28T23:27:29Z: hab_launch: Launcher exiting with code 0

[1]+  Done                    hab sup run
[67][default:/src:0]# ps fauxw | grep [r]edis
hab        569  0.2  0.0  42276  3764 pts/0    Sl   23:26   0:00 redis-server *:6379
```

Note that this is different from the normal behavior of killing hab-launcher in which all services are brought down:

```
[75][default:/src:130]# ps fauxw | grep [h]ab-
root       676  0.2  0.0  48056  6140 ?        Ssl  23:28   0:00 /hab/pkgs/core/hab-launcher/8479/20180917170345/bin/hab-launch run
root       679  0.7  0.1 209600 14524 ?        Sl   23:28   0:00  \_ /hab/pkgs/core/hab-sup/0.64.1/20180928012737/bin/hab-sup run
[76][default:/src:0]# ps fauxw | grep [r]edis
hab        796  0.2  0.0  42276  3844 ?        Sl   23:28   0:00  \_ redis-server *:6379
[77][default:/src:0]# kill 676
[78][default:/src:0]# ps fauxw | grep [r]edis
[79][default:/src:1]# ps fauxw | grep [h]ab-
[80][default:/src:1]# 
```

**Comment by christophermaier** [comment]
In this specific case, I've tracked the issue down to https://github.com/habitat-sh/habitat/pull/5404, but in general, I think it's an incompatibility between that and https://github.com/habitat-sh/habitat/pull/5445, which came afterwards. 

I'll continue to look into a fix, which may well be reverting #5404. In the meantime, you may be able to use [core/hab-launcher/7797/20180625172404](https://bldr.habitat.sh/#/pkgs/core/hab-launcher/7797/20180625172404) which was the last stable launcher before this code went in. I'm not able to reproduce what @stevendanna showed when I use that.

**Comment by stevendanna** [comment]
Thanks @christophermaier we can take a at pinning hab launcher with HAB_LAUNCH_BIN as we already have that pinned in our manifest.
